invoices/invoice_expiry_watcher.go

```diff
@@ -0,0 +1,178 @@
+	if err != nil {
+		log.Errorf("Error fetching invoices from the database: %v", err)
+	} else {
+		for k := 0; k < len(invoices); k++ {
+			ew.AddInvoice(hashes[k], &invoices[k])
+		}
+	}
+}
+
+// Start starts the the subscription handler and the main loop. Start() can
+// only be called once.
+func (ew *InvoiceExpiryWatcher) Start(cancelInvoice func(lntypes.Hash) error) {
+	ew.started.Do(func() {
+		ew.cancelInvoice = cancelInvoice
+		ew.wg.Add(1)
+		go ew.mainLoop()
+	})
+}
+
+// Stop cancels the invoice subscription and stops the expiry handler loop.
+func (ew *InvoiceExpiryWatcher) Stop() {
+	ew.stopped.Do(func() {
+		// Signal subscriptionHandler to quit and wait for it to return.
+		close(ew.quit)
+		ew.wg.Wait()
+	})
+}
+
+// AddInvoice adds a new invoice to the InvoiceExpiryWatcher. This won't check
+// if the invoice is already added and will only add invoices with ContractOpen
+// state.
+func (ew *InvoiceExpiryWatcher) AddInvoice(
+	paymentHash lntypes.Hash, invoice *channeldb.Invoice) {
+
+	if invoice.Terms.State != channeldb.ContractOpen {
+		log.Debugf("Invoice not added to expiry watcher: %v", invoice)
+		return
+	}
+
+	expiry := invoice.CreationDate.Add(
+		zpay32.DefaultInvoiceExpiry(invoice.Expiry))
+
+	log.Debugf("Adding invoice '%v' to expiry watcher, expiration: %v",
+		paymentHash, expiry)
+
+	ew.newInvoices <- &invoiceExpiry{
+		PaymentHash: paymentHash,
+		Expiry:      expiry,
+	}
+}
+
+// nextExpiry returns the duraton until we must wait for the next invoice to
+// expire. If there are no active invoices, then it simply returns max duration.
+func (ew *InvoiceExpiryWatcher) nextExpiry() time.Duration {
+	if !ew.expiryQueue.Empty() {
+		top := ew.expiryQueue.Top().(*invoiceExpiry)
+		return top.Expiry.Sub(ew.clock.Now())
+	}
+
+	return time.Duration(math.MaxInt64)
+}
+
+// cancelExpiredInvoices will cancel all expired invoices and removes them from
+// the expiry queue.
+func (ew *InvoiceExpiryWatcher) cancelExpiredInvoices() {
+	for !ew.expiryQueue.Empty() {
+		top := ew.expiryQueue.Top().(*invoiceExpiry)
+		if top.Expiry.Before(ew.clock.Now()) {
+			log.Infof("Cancelling expired invoice '%v', expiry: %v",
+				top.PaymentHash, top.Expiry)
+
+			if err := ew.cancelInvoice(top.PaymentHash); err != nil {
+				log.Errorf("Unable to cancel invoice: %v", top.PaymentHash)
+			}
+			ew.expiryQueue.Pop()
+		} else {
+			break
+		}
+	}
+}
+
+// mainLoop receives invoice events and handles cancellaton. Only new invoices
+// that are not already settled or canceled will be added to the queue.
+func (ew *InvoiceExpiryWatcher) mainLoop() {
+	defer ew.wg.Done()
+
+	for {
+		// Cancel any invoices that may have expired.
+		ew.cancelExpiredInvoices()
+
+		select {
+		case <-time.After(ew.nextExpiry()):
+			// Wait until the next invoice expires, then cancel expired invoices.
+			continue
+
+		case newInvoiceExpiry := <-ew.newInvoices:
+			ew.expiryQueue.Push(newInvoiceExpiry)
```

Comment:
I still wouldn't mind having those two events inside the main invoice registry main loop to reduce concurrency and any risks associated with that. But like this is ok too.

Reply:
I'd prefer keeping this separate as otherwise we introduce unrelated functionality to the subscriber loop.